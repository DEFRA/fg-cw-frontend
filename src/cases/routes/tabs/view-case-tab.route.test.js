import { load } from "cheerio";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import { createServer } from "../../../server/index.js";
import { findCaseTabUseCase } from "../../use-cases/find-case-tab.use-case.js";
import { viewCaseTabRoute } from "./view-case-tab.route.js";

vi.mock("../../use-cases/find-case-tab.use-case.js");

describe("viewCaseTabRoute", () => {
  let server;

  beforeEach(async () => {
    server = await createServer();
    server.route(viewCaseTabRoute);
  });

  afterEach(async () => {
    await server.stop();
  });

  it("renders tab view with correct data", async () => {
    const mockTabData = {
      caseId: "68c2ced6713f9a0233a1a1ea",
      caseRef: "10f-f72-e22",
      tabId: "case-details",
      banner: {
        title: {
          text: "",
          type: "string",
        },
        summary: {
          sbi: {
            label: "SBI",
            text: "106284736",
            type: "string",
          },
          reference: {
            label: "Reference",
            text: "10f-f72-e22",
            type: "string",
          },
          scheme: {
            label: "Scheme",
            text: "SFI",
            type: "string",
          },
          createdAt: {
            label: "Created At",
            text: "11 Sept 2025",
            type: "date",
          },
        },
      },
      links: [
        {
          id: "tasks",
          text: "Tasks",
          href: "/cases/68c2ced6713f9a0233a1a1ea",
          active: false,
        },
        {
          id: "case-details",
          text: "Case Details",
          href: "/cases/68c2ced6713f9a0233a1a1ea/case-details",
          active: true,
        },
        {
          id: "notes",
          text: "Notes",
          href: "/cases/68c2ced6713f9a0233a1a1ea/notes",
          active: false,
        },
        {
          id: "timeline",
          text: "Timeline",
          href: "/cases/68c2ced6713f9a0233a1a1ea/timeline",
          active: false,
        },
        {
          id: "agreements",
          text: "Agreements",
          href: "/cases/68c2ced6713f9a0233a1a1ea/agreements",
          active: false,
        },
      ],
      content: [
        {
          id: "title",
          component: "heading",
          text: "Application",
          level: 2,
          classes: "govuk-!-margin-top-6 govuk-!-margin-bottom-1",
        },
        {
          id: "submittedDate",
          component: "container",
          classes: "govuk-body",
          items: [
            {
              text: "submitted:",
              classes: "govuk-!-font-weight-regular",
            },
            {
              text: "11 Sept 2025",
              classes: "govuk-!-font-weight-bold",
            },
          ],
        },
        {
          id: "answers",
          component: "summary-list",
          title: "Answers",
          type: "object",
          rows: [
            {
              text: "SFI",
              type: "string",
              label: "Scheme",
            },
            {
              text: 2025,
              type: "number",
              label: "Year",
            },
            {
              text: "Yes",
              type: "boolean",
              label: "Has checked land is up to date?",
            },
            {
              text: "NO_LONGER_REQUIRED",
              type: "string",
              label: "Agreement Name",
            },
          ],
        },
      ],
    };

    findCaseTabUseCase.mockResolvedValue(mockTabData);

    const { statusCode, result } = await server.inject({
      method: "GET",
      url: "/cases/case-123/case-details",
    });

    expect(statusCode).toBe(200);
    expect(findCaseTabUseCase).toHaveBeenCalledWith("case-123", "case-details");

    const $ = load(result);
    const view = $("#main-content").html();

    expect(view).toMatchSnapshot();
  });

  it("handles different tab types", async () => {
    const mockTimelineData = {
      _id: "case-456",
      caseRef: "TIM-2024-002",
      tabId: "timeline",
      banner: {
        title: {
          text: "",
          type: "string",
        },
        summary: {
          sbi: {
            label: "SBI",
            text: "106284736",
            type: "string",
          },
          reference: {
            label: "Reference",
            text: "10f-f72-e22",
            type: "string",
          },
          scheme: {
            label: "Scheme",
            text: "SFI",
            type: "string",
          },
          createdAt: {
            label: "Created At",
            text: "11 Sept 2025",
            type: "date",
          },
        },
      },
      links: [
        {
          id: "tasks",
          text: "Tasks",
          href: "/cases/68c2ced6713f9a0233a1a1ea",
          active: false,
        },
        {
          id: "case-details",
          text: "Case Details",
          href: "/cases/68c2ced6713f9a0233a1a1ea/case-details",
          active: true,
        },
        {
          id: "notes",
          text: "Notes",
          href: "/cases/68c2ced6713f9a0233a1a1ea/notes",
          active: false,
        },
        {
          id: "timeline",
          text: "Timeline",
          href: "/cases/68c2ced6713f9a0233a1a1ea/timeline",
          active: false,
        },
        {
          id: "agreements",
          text: "Agreements",
          href: "/cases/68c2ced6713f9a0233a1a1ea/agreements",
          active: false,
        },
      ],
      content: [
        {
          id: "title",
          component: "heading",
          text: "Timeline",
          level: 2,
          classes: "govuk-!-margin-top-6 govuk-!-margin-bottom-1",
        },
        {
          id: "submittedDate",
          component: "container",
          classes: "govuk-body",
          items: [
            {
              text: "submitted:",
              classes: "govuk-!-font-weight-regular",
            },
            {
              text: "11 Sept 2025",
              classes: "govuk-!-font-weight-bold",
            },
          ],
        },
      ],
    };

    findCaseTabUseCase.mockResolvedValue(mockTimelineData);

    const { statusCode, result } = await server.inject({
      method: "GET",
      url: "/cases/case-456/timeline",
    });

    expect(statusCode).toBe(200);
    expect(findCaseTabUseCase).toHaveBeenCalledWith("case-456", "timeline");

    const $ = load(result);
    const view = $("#main-content").html();

    expect(view).toMatchSnapshot();
  });

  it("handles use case returning null", async () => {
    findCaseTabUseCase.mockResolvedValue(null);

    const response = await server.inject({
      method: "GET",
      url: "/cases/nonexistent/tab",
    });

    expect(response.statusCode).toBe(500);
    expect(findCaseTabUseCase).toHaveBeenCalledWith("nonexistent", "tab");
  });

  it("handles use case errors", async () => {
    const useCaseError = new Error("Database connection failed");
    findCaseTabUseCase.mockRejectedValue(useCaseError);

    const response = await server.inject({
      method: "GET",
      url: "/cases/case-error/tab-error",
    });

    expect(response.statusCode).toBe(500);
    expect(findCaseTabUseCase).toHaveBeenCalledWith("case-error", "tab-error");
  });

  it("extracts parameters correctly from URL", async () => {
    const mockTabData = {
      _id: "param-test-case",
      caseRef: "PARAM-2024-001",
      links: [
        { id: "tasks", text: "Tasks", href: "/cases/case-456" },
        {
          id: "agreements",
          text: "Agreements",
          href: "/cases/case-456/timeline",
        },
      ],
    };

    findCaseTabUseCase.mockResolvedValue(mockTabData);

    const { statusCode, result } = await server.inject({
      method: "GET",
      url: "/cases/param-test-case/agreements",
    });

    expect(statusCode).toBe(200);
    expect(findCaseTabUseCase).toHaveBeenCalledWith(
      "param-test-case",
      "agreements",
    );

    const $ = load(result);
    const view = $("#main-content").html();

    expect(view).toMatchSnapshot();
  });
});
