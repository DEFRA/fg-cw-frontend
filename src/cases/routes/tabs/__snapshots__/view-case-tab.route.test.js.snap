// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`viewCaseTabRoute > extracts parameters correctly from URL 1`] = `
"
  
<div class="govuk-!-margin-bottom-6 app-case-banner">
  <div class="govuk-width-container">
    <a href="/cases" class="govuk-back-link">Back to cases</a>

    <h1 class="govuk-heading-m">
      Valley Farm
    </h1>

    <div>
      <div class="app-case-row">
      </div>
    </div>
  </div>
</div>


  
  <div class="govuk-service-navigation" data-module="govuk-service-navigation">
      <div class="govuk-width-container">

    <div class="govuk-service-navigation__container">
      

      
        <nav aria-label="Menu" class="govuk-service-navigation__wrapper">
          <button type="button" class="govuk-service-navigation__toggle govuk-js-service-navigation-toggle" aria-controls="navigation" hidden="">
            Menu
          </button>

          <ul class="govuk-service-navigation__list" id="navigation">

            
              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/case-456">
                                    
Tasks
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item govuk-service-navigation__item--active">
                  <a class="govuk-service-navigation__link" href="/cases/case-456/timeline" aria-current="true">
                                    
                  <strong class="govuk-service-navigation__active-fallback">Agreements</strong>

                  </a>
              </li>

            </ul>
        </nav>
    </div>

    </div>

  </div>


  
  <p class="govuk-body">No content available.</p>

        "
`;

exports[`viewCaseTabRoute > handles different tab types 1`] = `
"
  
<div class="govuk-!-margin-bottom-6 app-case-banner">
  <div class="govuk-width-container">
    <a href="/cases" class="govuk-back-link">Back to cases</a>

    <h1 class="govuk-heading-m">
      Valley Farm
    </h1>

    <div>
      <div class="app-case-row">
            <div class="app-case-cell">
              <p>
                <strong class="govuk-!-margin-right-1">SBI:</strong>
                106284736
              </p>
            </div>
            <div class="app-case-cell">
              <p>
                <strong class="govuk-!-margin-right-1">Reference:</strong>
                10f-f72-e22
              </p>
            </div>
            <div class="app-case-cell">
              <p>
                <strong class="govuk-!-margin-right-1">Scheme:</strong>
                SFI
              </p>
            </div>
            <div class="app-case-cell">
              <p>
                <strong class="govuk-!-margin-right-1">Created At:</strong>
                11 Sept 2025
              </p>
            </div>
      </div>
    </div>
  </div>
</div>


  
  <div class="govuk-service-navigation" data-module="govuk-service-navigation">
      <div class="govuk-width-container">

    <div class="govuk-service-navigation__container">
      

      
        <nav aria-label="Menu" class="govuk-service-navigation__wrapper">
          <button type="button" class="govuk-service-navigation__toggle govuk-js-service-navigation-toggle" aria-controls="navigation" hidden="">
            Menu
          </button>

          <ul class="govuk-service-navigation__list" id="navigation">

            
              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68c2ced6713f9a0233a1a1ea">
                                    
Tasks
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68c2ced6713f9a0233a1a1ea/case-details">
                                    
Case Details
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68c2ced6713f9a0233a1a1ea/notes">
                                    
Notes
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item govuk-service-navigation__item--active">
                  <a class="govuk-service-navigation__link" href="/cases/68c2ced6713f9a0233a1a1ea/timeline" aria-current="true">
                                    
                  <strong class="govuk-service-navigation__active-fallback">Timeline</strong>

                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68c2ced6713f9a0233a1a1ea/agreements">
                                    
Agreements
                  </a>
              </li>

            </ul>
        </nav>
    </div>

    </div>

  </div>


  
    
    
  <h2 class="govuk-heading-l  govuk-!-margin-top-6 govuk-!-margin-bottom-1">
    Timeline
  </h2>


    
    
  <div id="submittedDate" class="defra-container govuk-body">
      
    
  <span class="govuk-!-font-weight-regular">submitted:</span>


      
    
  <span class="govuk-!-font-weight-bold">11 Sept 2025</span>


  </div>



        "
`;

exports[`viewCaseTabRoute > renders tab view with correct data 1`] = `
"
  
<div class="govuk-!-margin-bottom-6 app-case-banner">
  <div class="govuk-width-container">
    <a href="/cases" class="govuk-back-link">Back to cases</a>

    <h1 class="govuk-heading-m">
      Valley Farm
    </h1>

    <div>
      <div class="app-case-row">
            <div class="app-case-cell">
              <p>
                <strong class="govuk-!-margin-right-1">SBI:</strong>
                106284736
              </p>
            </div>
            <div class="app-case-cell">
              <p>
                <strong class="govuk-!-margin-right-1">Reference:</strong>
                10f-f72-e22
              </p>
            </div>
            <div class="app-case-cell">
              <p>
                <strong class="govuk-!-margin-right-1">Scheme:</strong>
                SFI
              </p>
            </div>
            <div class="app-case-cell">
              <p>
                <strong class="govuk-!-margin-right-1">Created At:</strong>
                11 Sept 2025
              </p>
            </div>
      </div>
    </div>
  </div>
</div>


  
  <div class="govuk-service-navigation" data-module="govuk-service-navigation">
      <div class="govuk-width-container">

    <div class="govuk-service-navigation__container">
      

      
        <nav aria-label="Menu" class="govuk-service-navigation__wrapper">
          <button type="button" class="govuk-service-navigation__toggle govuk-js-service-navigation-toggle" aria-controls="navigation" hidden="">
            Menu
          </button>

          <ul class="govuk-service-navigation__list" id="navigation">

            
              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68c2ced6713f9a0233a1a1ea">
                                    
Tasks
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item govuk-service-navigation__item--active">
                  <a class="govuk-service-navigation__link" href="/cases/68c2ced6713f9a0233a1a1ea/case-details" aria-current="true">
                                    
                  <strong class="govuk-service-navigation__active-fallback">Case Details</strong>

                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68c2ced6713f9a0233a1a1ea/notes">
                                    
Notes
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68c2ced6713f9a0233a1a1ea/timeline">
                                    
Timeline
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68c2ced6713f9a0233a1a1ea/agreements">
                                    
Agreements
                  </a>
              </li>

            </ul>
        </nav>
    </div>

    </div>

  </div>


  
    
    
  <h2 class="govuk-heading-l  govuk-!-margin-top-6 govuk-!-margin-bottom-1">
    Application
  </h2>


    
    
  <div id="submittedDate" class="defra-container govuk-body">
      
    
  <span class="govuk-!-font-weight-regular">submitted:</span>


      
    
  <span class="govuk-!-font-weight-bold">11 Sept 2025</span>


  </div>


    
    

     
  <h3 class="govuk-heading-m ">
    Answers
  </h3>

   <dl class="govuk-summary-list">
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key">
      Scheme
    </dt>
    <dd class="govuk-summary-list__value">
      <span>SFI</span>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key">
      Year
    </dt>
    <dd class="govuk-summary-list__value">
      <span>2025</span>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key">
      Has checked land is up to date?
    </dt>
    <dd class="govuk-summary-list__value">
      <span>Yes</span>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key">
      Agreement Name
    </dt>
    <dd class="govuk-summary-list__value">
      <span>NO_LONGER_REQUIRED</span>
    </dd>
  </div>
</dl>



        "
`;
